File: examples/pizzas_go/domain/add_topping.go
package domain

import "time"

type AddTopping struct {
	PizzaId string `json:"pizza_id"`
	Name string `json:"name"`
	Amount int64 `json:"amount"`
}

func (c AddTopping) CommandName() string { return "AddTopping" }

func (c AddTopping) Execute(repo PizzaRepository) (*Pizza, *AddedTopping, error) {
	existing, err := repo.Find(c.PizzaId)
	if err != nil {
		return nil, nil, err
	}
	if existing == nil {
		return nil, nil, fmt.Errorf("Pizza not found: %s", c.PizzaId)
	}
	existing.Name = c.Name
	existing.Amount = c.Amount
	existing.UpdatedAt = time.Now()
	if err := existing.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(existing); err != nil {
		return nil, nil, err
	}
	event := AddedTopping{
		AggregateID: existing.ID,
		PizzaId: c.PizzaId,
		Name: c.Name,
		Amount: c.Amount,
		OccurredAt: time.Now(),
	}
	return existing, &event, nil
}

File: examples/pizzas_static/lib/pizzas_domain/pizza/commands/add_topping.rb
module PizzasDomain
  class Pizza
    module Commands
      class AddTopping
        include PizzasDomain::Runtime::Command
        emits "AddedTopping"

        attr_reader :pizza_id
        attr_reader :name
        attr_reader :amount

        def initialize(
          pizza_id: nil,
          name: nil,
          amount: nil
        )
          @pizza_id = pizza_id
          @name = name
          @amount = amount
        end

        def call
          existing = repository.find(pizza_id)
          if existing
            Pizza.new(
              id: existing.id,
              name: name,
              description: existing.description,
              toppings: existing.toppings + [Topping.new(name: name, amount: amount)]
            )
          else
            raise Hecks::Error, "Pizza not found: #{pizza_id}"
          end
        end
      end
    end
  end
end
